Encryption.md

```diff
@@ -75,55 +75,126 @@ written to the output stream.
-The pages are encrypted with AES-CTR, where the IVs (16 bytes) are written before each 
-ciphertext.
+Thrift modules are encrypted with AES-GCM, as described above. The pages are encrypted by an AES-CTR
+cipher without padding. The nonces are generated and written before each ciphertext, with the nonce length of 16 bytes 
+(128 bits). At least 12 bytes out of 16 are constructed with an RBG.
+
+```c
+struct AesGcmV1 {
+  /** Retrieval metadata of AAD prefix used for encryption of pages and structures **/
+  1: optional binary aad_metadata
+}
+
+struct AesGcmCtrV1 {
+  /** Retrieval metadata of AAD prefix used for encryption of structures **/
+  1: optional binary aad_metadata
+}
+
+union EncryptionAlgorithm {
+  1: AesGcmV1 AES_GCM_V1
+  2: AesGcmCtrV1 AES_GCM_CTR_V1
+}
+```
+
+### Additional Authenticated Data
+
+The AES GCM cipher in the basic mode protects against byte replacement inside a ciphertext, 
+but can't prevent replacement of one ciphertext with another (encrypted with the same key). 
+This can be solved by the AAD mode, using different AADs for different ciphertexts. Parquet 
+modular encryption uses the AAD mode to protect against swapping ciphertext modules inside a 
+file, between files - or against swapping full files (for example, replacement of a file with an 
+old version). Obviously, the AAD must reflect the identity of a file and of the modules inside 
+the file. 
+
+Parquet constructs a module AAD from two components: an optional AAD prefix - a string provided 
+by the user for the file, and an AAD suffix, built internally for each encrypted module inside 
+the file. The AAD prefix should reflect the target identity that helps to detect file swapping - 
+for example, table name and version or date / timestamp (e.g., "employees_23_05_2018"). The 
+AAD suffix reflects the internal identity of modules inside the file, which for example prevents 
+replacement of column pages in row group 0 by pages from the same column in row group 1. The module 
+AAD is a direct concatenation of the prefix and suffix parts. 
+
+#### AAD prefix
+
+A file writer passes an AAD prefix string upon file creation, that allows to differentate it from other
+files and datasets. The reader should know this string apriori, or should be able to retrieve and verify
+it, using a convention accepted in the organization.
+For example, if the convention is to use a table name and date as the AAD prefix, a 
+writer of a May 23 version of an employee table will use "employees_23_05_2018" as an AAD 
+prefix (and as the file name). The reader that needs to process the May 23 table, knows the AAD must be 
+"employees_23_05_2018", without relying on the name of the encrypted file.
 
+The `AesGcmV1` and `AesGcmCtrV1` structures contain an optional `aad_metadata` field that can 
+be used by a reader to retrieve the AAD prefix string used for file enciphering. The maximal allowed
+length of `aad_metadata` is 512 bytes. If organization works with a number of different 
```

Comment:
Why is this limited to 512 bytes? Is this a restriction from AES-GCM?

Reply:
No good reason, actually, except for an effort to limit an input into this free-form byte array. If this is not the right way/place to do it, I'll remove this text.

Reply:
Let's remove it. No need for arbitrary limitations that could require work-arounds in the future.